[RFC] abstract clock subsystem: counter syntonization

fgrandel opened the pull request:
Proposes a well-defined intermediate concept of scalar syntonized nanosecond resolution time with overflow protection above low-level counters/cycles/ticks and below higher level time abstractions (timescales, calenders, etc.). Details see the discussion below.

Currently this happens in the net subsystem but in principle this can and may be promoted to something more generic as it is an implementation of layers 1 (counter) and 2 (syntonization) of the proposed [clock subsystem RFC](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731) which is itself relevant to the POSIX roadmap.

The infrastructure discussed in this RFC will first be used for [TSCH TDMA](https://github.com/zephyrproject-rtos/zephyr/issues/50336).

fgrandel commented:
@rlubos @jciupis @edmont relevant to you in the context of #59245

fgrandel commented:
@cfriedt relevant to you in the context of POSIX clocks

kartben commented on include/zephyr/sys/ktime.h:
```diff
@@ -0,0 +1,134 @@
+#define USEC_PER_SEC ((USEC_PER_MSEC) * (MSEC_PER_SEC))
+
+/* number of nanoseconds per second */
+#define NSEC_PER_SEC ((NSEC_PER_USEC) * (USEC_PER_MSEC) * (MSEC_PER_SEC))
```
Use javadoc style comments /** ... */

fgrandel replied:
@kartben I forgot the @cond INTERNAL_HIDDEN wrapper around these constants (now added). I also made them /** ... */ comments now. Not sure whether this is still required for hidden constants, though.

@andyross The reason is that you were not much in favor of exposing those constants too much - at least not as a well-published API that incentivates further proliferation of their use.

kartben replied:
Ya if they're hidden you're fine dropping the Javadoc style comments. That would actually help make it even clearer they're not "API" (which I didn't know if they were when originally commenting)

kartben replied:
Ya if they're hidden you're fine dropping the Javadoc style comments. That would actually help make it even clearer they're not "API" (which I didn't know if they were when originally commenting)

fgrandel replied:
Makes sense. Returned to inline /* ... */ comments. :-)

kartben commented on include/zephyr/sys/ktime.h:
```diff
@@ -0,0 +1,134 @@
+/** The largest positive number of seconds that can be safely represented by ktime_t */
+#define KTIME_SEC_MAX (KTIME_MAX / NSEC_PER_SEC)
+
+/** The largest negative number of seconds that can be safely represented by ktime_t */
```
...smallest...

fgrandel replied:
Fixed.

kartben commented on include/zephyr/sys/ktime.h:
```diff
@@ -0,0 +1,134 @@
+/** The largest positive time value that can be represented by ktime_t */
+#define KTIME_MAX INT64_MAX
+
+/** The largest negative time value that can be represented by ktime_t */
```
... smallest ...

fgrandel replied:
Fixed.

andyross reviewed (commented):
No real objections to the typedef, but it seems a little nonorthogonal?  Right now kernel time is in ticks, because a "tick" is what kernel time is measured in by definition.  Conversion of ticks to ns (both numerically and via the existing timeout mechanism) is already provided for anyone who needs it, and the utilities to do so are uniformly 64 bit (there are 32 bit converters for smaller units, but not ns).

Having a separate "kind" of time seems more confusing than helpful to me, honestly.  There aren't any APIs that use this yet.  What's the end goal here?

cfriedt commented:
@andyross - I would almost suggest a 1:1 with Florian. I had a few broader picture discussions with him about some hi-res timer / clock things and was sold on it. Aside, those old tickets might actually get closed soon!

andyross commented:
Sure.  There's just a long graveyard of attempts in this space, and I'd much rather we see an implementation/prototype arrive before making any changes to the public API.  In particular an attempt to tie kernel timekeeping to a real unit is something we spilled a lot of blood trying to get away from, given that so many of our platforms use counters that don't convert cleanly into decimal units.

fgrandel commented:
> I'd much rather we see an implementation/prototype arrive before making any changes to the public API

@andyross : 100% agreed - the onus is on me to prove that it makes sense to decouple certain uses of time from existing kernel concepts (ticks/cycles) and that ktime_t is the right candidate to do so. I agree with @cfriedt that a F2F would be helpful.

In any case I provide some more thoughts in writing - depending on your communication preference **no need to read the following fine print, though, if you prefer F2F anyway...**

> Conversion of ticks to ns (both numerically and via the existing timeout mechanism) is already provided for anyone who needs it

The new type is not meant to represent kernel time "on the fly" but to represent time in structures and variables independently of the underlying clock source which more often than not will not be the kernel cycle counter.

> given that so many of our platforms use counters that don't convert cleanly into decimal units.

Rounding from/to the ns domain is dealt with explicitly in the proposed algorithms as it is - as you say - a practical problem - albeit an unavoidable one when working with hybrid clock sources which is at the core of the ktime_t proposal (see below for details).

> There aren't any APIs that use this yet.

@cfriedt , @andyross : Do you agree that **the [waiting PR](https://github.com/zephyrproject-rtos/zephyr/pull/60401) justifies ktime_t at least for the network subsystem**? Would you prefer that we "hide" ktime_t somewhere in the network subsystem then until we provide PRs in the POSIX/libc/TSCH area using that type in a broader sense?

IMO the ktime_t concept has been very useful to refactor time across the network subsystem (including but not limited to the IEEE 802.15.4 subsystem):
 * Attempts to use generic types to represent time in the network subsystem failed in the past as a precise common specification of time was never provided. Fields representing time were used inconsistently within the same subsystem and across drivers due to lack of a well-defined concept of high-resolution scalar time.
 * Standardizing around kernel cycles in net_pkt or the radio API wouldn't work as kernel cycles are not guaranteed to be high resolution enough to represent network time plus the clock source is almost never kernel time (but comparison to / synchronization with kernel time is still needed).
 * Other existing notions of time would be inadequate to represent time in the network subsystem as well. The rationale is given in the description of ktime_t for all such types representing time in Zephyr so far.

> What's the end goal here?

As soon as this PR is accepted that code will be added to the code base immediately introducing ktime_t as network time.

But that alone would not justify introducing ktime_t in such an exposed place. We all need to be 100% convinced that at least the following features make sense in Zephyr in the short to mid term, should be decoupled from kernel time and use ktime_t-based intermediate clock sources instead:
 1. high-resolution POSIX clocks and libc time sharing a common adjusted and drift compensated low-power (tickless) hybrid RTC/high-res counter clock source which may or may not be (partially) based on the kernel timer and can be configured in DT to use other peripherals instead.
 2. hybrid nanosecond resolution clock sources for TDMA / timed TX/RX protocols in the IEEE 802.15.4 stack (namely CSL and TSCH) based on the combination of a low-power, always-on RTC and a radio timer providing high resolution during RX/TX windows which will usually not involve kernel time at all
 3. synchronization/compensation of the main internal clock source (and thereby POSIX/libc clocks) with low-frequency, high-precision "pulse-per-second" devices like always-on RTCs, NTP, PTP, GPS, clocks derived from TDMA radio or industrial automation protocols or even atomic clock modules.

These use cases have in common:
* They all rely on **"hybrid" clock sources** combining at least two hardware peripherals or drivers of which the kernel timer is only one (if used at all).
* They **share a couple of algorithms** (drift compensation, leap second support, epoch offsets, etc.) which should be re-usable across all of them as a shared clock subsystem and which cannot sensibly be based on ticks/cycles for sufficient precision but provably require a well defined concept of scalar nanosecond resolution time like ktime_t.

**Next steps** that follow immediately will be:
 * I'll provide PRs for the basics of a [clock subsystem](https://github.com/zephyrproject-rtos/zephyr/issues/19030) with support for generic hybrid clock sources, drift adjustment + well-defined timescales (this is when ktime_t would have to be moved from the network subsystem to something shared).
 * @cfriedt , if I understood you correctly you'll then be able to use that clock subsystem to solve the current layering problem in POSIX and libc plus introduce a DT-based configuration approach for the hybrid POSIX/libc clock source.
 * At the same time I'll start to base TSCH timing on the same common clock subsystem.

keith-packard commented:
> > @kartben The build suddenly fails with "API_DEFINE: do not specify a non-Zephyr API for libc File:include/zephyr/sys/ktime.h". Any idea what that means? The only thing that changed is the comments.
> 
> @keith-packard would know :) See #60487

Whoa. Looks like that new check is busted. Lemme go sort it out.

cfriedt commented:
> tl;dr: No -1 from me. But... I really think this needs to wait on some major subsystem work using it to land to see what the impact is going to be.

I wonder if it's possible to isolate things to a testsuite or sample application first, get some of these concepts upstream that way without it being public API, and then create a separate PR that essentially flips a switch (moves the headers into public API and demonstrates the MVP use case). That way, it's still a 2-way door.

A testsuite would kind of be ideal because we could test things on various pieces of hw, with e.g. a 32k timer.

fgrandel commented:
> I really think this needs to wait on some major subsystem work using it to land to see what the impact is going to be.

@andyross I think you're right. We need a deeper understanding of how this type is going to be embedded in a larger subsystem based on real code. I have the same concerns re aliasing and bad assumptions about / abuse of pseudo-precise nanosecond values. These problems need to be addressed right away - not later on. I propose I write more code and draw some pictures, add it all into this PR and then we continue the discussion.

> I wonder if it's possible to isolate things to a testsuite or sample application first, get some of these concepts upstream that way without it being public API

@cfriedt Yes, this goes in the direction I proposed above when I said that we should probably hide this typedef in the network subsystem while it's only used there. I believe it is provably useful there - it's a big step forward, even if it never makes its way out of the network subsystem. Before we use the concept in POSIX clocks or libc, @andyross 's concerns need to be addressed anyway.

> In particular, the network layer is in fact a pretty bad citizen here precisely because it's not tick-aware!

@andyross Do you propose to base TX/RX timestamps, timed TX and RX windows on opaque counter cycle values? I actually considered this, but so far I assumed that this would be a bad idea. None of the standards relevant in the area (IP derivatives, gPTP, IEEE 802.15.4, Thread, Bluetooth) use counter values. They all use nanosecond values, microsecond values or timespec-like structs with well defined tolerances concerning drift and jitter (which usually excludes 32k oscillators anyway). Then these standards deal explicitly with aliasing problems but hide them behind MAC (L2) or PHY (L1) APIs. Therefore I currently assume that using nanoseconds + timespec as the common denominator for network APIs is the right approach even in the long term. But I might be blind here because I read these standards all day - so what is your opinion about this? It's important because it influences the clock subsystem design quite a bit.

keith-packard commented:
I have watched other projects adopt a general 64-bit nanosecond time mechanism with good results; it avoids numerous errors where scaling is performed incorrectly or lost entirely. Plus it encourages the use of 64-bit types instead of various kludges to try and fit times into 32-bit values. As long as the model exposes the resolution of every time source, it seems like the resulting programming model is generally easier to use correctly.

andyross commented:
For clarity here: I'm not saying an API that exposes fine-grained monotonic time is a bad idea.  I'm not saying it doesn't belong in Zephyr somewhere.  I'm saying it's a poor fit for our hardware and that it almost certainly doesn't belong in the kernel as the "official" way we keep time.

Again, we spent **years** dealing with test code that would naively compute timeouts in milliseconds in an obvious way and then fall over mysteriously on Nordic because something happened too early or took too long.  And the solution was to say "No, don't give me an integer.  Instantiate a timeout once and the system will ensure it's an opaque struct so you don't shoot your foot trying to do more math on it."

If we could assume all systems had a MHz-scale timekeeping device, my answer would be different.  But... we can't.

andyross commented:
> @andyross Do you propose to base TX/RX timestamps, timed TX and RX windows on opaque counter cycle values? I actually considered this, but so far I assumed that this would be a bad idea:

To be blunt: if you're basing it on something other than ticks, regardless of what your API says, you probably have bugs somewhere.  The result of `k_uptime_ticks()` is the official kernel monotonic timestamp source.

fgrandel commented:
> it almost certainly doesn't belong in the kernel as the "official" way we keep time.

True. You're right, that I'm not proposing to replace the kernel's existing cycle/tick based timekeeping method, it should remain untouched. Let me draw some pictures (I'm already on it) and then let's talk over it.

andyross commented:
FWIW, because @npitre just pushed a new version and I remembered it, see #60232 for a very orthogonal take on what amounts to the same problem.  That goes even further than the current API does at forcing the API users into opaque usage of (internally maintained) tick counts and discouraging dangerous application-space math.

fgrandel commented:
The more I learn about syntonization, the more I'm confident that framing the discussion about ns vs. cycle representation as being mutually exclusive is an undue over-simplification which is why I personally never claimed that one concept is generally "better" than the other, see the precise divide between cycle- vs. ns-based time representation in the attached concept map - proposed additions to Zephyr in red). Sorry for the ugly design, I had no time to make it beautiful.

[zephyr-clock-subsystem-concept-map.pdf](https://github.com/zephyrproject-rtos/zephyr/files/12138828/zephyr-clock-subsystem-concept-map.pdf)

**Please don't engage in a discussion about names or missing API details in that PDF, it is not meant as an API specification but as a precise representation of architectural layers** with reference to existing SoCs, peripherals, driver APIs and protocols I personally know. I consider this *team work in progress* to be constructively evolved along diverse input from several perspectives (including prior discussion in this PR).

**I therefore propose that we re-focus our discussion here on constructively shaping one specific state-of-the-art clock subsystem layer (namely [syntonization](https://www.researchgate.net/profile/Bengt-Olsson-2/publication/329706705/figure/fig1/AS:704742024019968@1545035000121/SYNCHRONIZED-VS-SYNTONIZED-CLOCKS-BOTH-PROGRESS-AT-THE-SAME-RATE-AS-THEIR-MASTERS-BUT-A.ppm) as opposed to counter, timescale and calendar layers) in Zephyr.**

Higher layers are [generally discussed elsewhere](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731). Discussing details of layers below and above the syntonization layer can and should IMO be reserved to other PRs. Replacing, evolving or even discussing existing kernel timer/counter concepts was never intended.

To summarize results from my pov wrt to the counter vs syntonization divide so far:

* **Cycles/tick representation is adequate when time spans or points in time wrt a specific counter need to be represented** (this includes kernel scheduling and @npitre 's contribution as @andyross has correctly pointed out but also representation of any other local counter-style device). Some of the arguments given here were highly relevant to the counter layer including but not limited to kernel ticks/cycles but not to syntonization requirements.

* **Nanosecond representation is adequate when any number of local or distributed hardware counters need to be syntonized "as close as possible" to a precise pulse-per-second (PPS) or remote reference clock** (NTP, PTP, local PPS clock sources, precision control applications, timed TX/RX and slotted network protocols). PPS or remote reference time will almost always be low-frequency (1Hz or lower) with tick jitter orders of magnitude worse than timestamp precision. Tick counter representation is systematically inadequate for reference clock devices even on [hardware driver level](https://datatracker.ietf.org/doc/html/rfc2783) . Network subsystem timestamps and timed TX/RX are one but not the only application of high-precision cross-counter syntonized time representation in Zephyr.

Some additional comments that hopefully help to understand the syntonization layer of the attached concept map:

It is easy to show that together with a deterministic rounding method, typical syntonization algorithms define an injective (non-surjective, non-invertible) functional relation between low-frequency, high precision nano- or microsecond reference timestamps and typcial local crystal oscillator based counter values with frequencies anywhere below 1GHz and well defined maximum phase jitter and frequency drift/wander. It is therefore easy to check whether a given local counter peripheral is sufficiently precise and accurate for any given syntonization requirement. Typical state-of-the-art reference clocks operate in the range of ns (PTP, GPS, atomic clocks) to us (TDMA, BLE, radio timers, PROFIBUS, ...) precision. Arguing that such precision is not relevant for "our" Zephyr hardware is therefore invalid IMHO. For details how this works across devices refer to published syntonization algorithms like IEEE 1588 (PTP) or IEEE 802.15.4 (TSCH TDMA, CSL).

The same level of precision cannot be achieved converting between time represented in terms of sub-GHz counter values (including but not limited to kernel ticks/cycles). Upverting to counters with higher resolution or better accuracy (e.g. radio timers) causes undue loss of precision. Representing distributed, syntonized reference time in terms of local hardware cycle counters therefore systematically leads to unnecessarily and inacceptable syntonization error. Choosing the "best" local counter as reference clock is no solution either as any locally unknown remote counter might be even "better".

That said, syntonizing clocks based on "virtual" arbitrary precision cycle counter values is possible, adequate and recommended even for reference clocks. Reference clocks can be conceptualized as 1GHz or 1MHz virtual counters without loss of generality. Peter Bigot's [nice and simple time synchronization utilities](https://docs.zephyrproject.org/latest/kernel/timeutil.html#time-scale-synchronization) implement one of many possible counter-based syntonization algorithms. It is not the most common nor the most precise syntonization algorithm. But that just shows the need for ["pluggability"](https://github.com/zephyrproject-rtos/zephyr/issues/19030#issuecomment-1597226731) of syntonization algorithms. In motion control, automation, PTP or TDMA applications, more general FIR/PI-style control loops are common.

To summarize: Syntonization requires probabilistic and statistical modeling. Concepts from digital control theory, especially phase or frequency locked loops (PLL/FLL), are commonly applied. Syntonization is usually modeled as a discrete time or z-domain feedback control loop (inclucing e.g. digital FIRs or PI controllers) with precise statistical and probabilistic properties that cannot adequately be made "precise" in terms of well-defined counter registers.

Commit:
kernel: timer/timeout: make timer instance aware

Introduces a dependency to a specific timout API into the timer
structure. This allows multiplexed timers to be defined against
arbitrary tick counter sources.

Also fixes a minor locking issue that would allow concurrent read/write
access on timeout nodes due to insufficient locking.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: introduce network uptime reference API

Introduces the API for a syntonized network uptime reference including
the underlying hybrid (sleep/high-resolution) counter.

Also prepares the IEEE 802.15.4 driver API as a client of this API.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
@rlubos @vaishnavachath Please do not waste your time looking into this yet unless you're personally interested.

@cfriedt @peter-mitsis You already know this dev branch - nothing fundamentally new since you looked into it.

@andyross I just upgraded the branch behind this RFC to early ongoing development work. No need to look closely as this will be split into several specialized PRs for actual review once it reaches production quality. I just want to give you an optional (!) ongoing preview of the "big picture" in the context of this RFC (see net_time.h) with some sample code for one SoC series (TI CC13/26xx) *iif* you want to have a look. Please ignore if you don't have the time.

Preview Content:
* Changes to net_time.h that explain how I propose to base syntonized time on top of tick-based counters. This corresponds to what has been discussed in the comments above. (Tick != kernel tick - just a synonym for an arbitrary counter's clock cycle in the context of this RFC.)
* Proposal to re-use the existing tick-based timer/timeout multiplexing code in the kernel on top of arbitrary counter drivers.
* Proposal to make the existing timer/timeout code more deterministic in the context of high-precision syntonized uptime by non-intrusively locking the elapsed() value. This corresponds to what has been discussed on Discord.
* Proposal to combine a low-power sleep counter with a high-resolution counter into a single non-wrapping high-resolution uptime tick space hiding away the complexities power management and hardware-based counter synchronization. From an API perspective this is done in a vendor-neutral way so that different implementation patterns based on bare metal programming and/or building upon existing drivers from different categories is possible.

Todo:
* Peter Bigot's syntonization function (or an alternative mean deviation syntonization algorithm) is not yet hooked into the implementation but I think it's easy to see how branching into any pluggable algorithm can be done.
* Overflow protection has not yet been implemented for the CC13/26xx peripheral.
* The details of counter synchronization and hybrid alarm programming are probably still buggy. I need to test this more.
* Pulse-per-second devices will be needed to synchronize sleep and hi-res counter drivers in a vendor-agnostic way. For CC13/26xx there's no need for a PPS driver as TI's radio driver internally synchronizes the two counters of interest (as does Nordic's).
* In the case of the CC13/26xx syntonization is done on hardware level, so no need to update running timers after a syntonization event. In the general case we may need a way to update active timeouts. But this is nice-to have as in practice syntonization is rarely done while timers are running.

@cfriedt If we want to use this as a basis for syntonized (PTP/NTP adjustable) POSIX/libc clocks it'll have to be promoted from the net subsys to something more general. A trivial fallback implementation based on kernel timers can be provided. I propose that we test this in the TDMA context for a while. If it turns out to be useful there we can easily promote it. Conceptually this is just an implementation of the proposed generic clock framework layers 1 (counter clock source) and 2 (syntonization).

fgrandel commented:
Here comes an example from my dev branch that shows how the different time abstractions work together:

```c
	case IEEE802154_TX_MODE_TXTIME: {
		k_timepoint_t tx_timepoint;
		ratmr_t rat_tx_time;
		net_time_t tx_time;

		/* Convert the network timestamp to syntonized network uptime. */
		tx_time = net_ptp_time_to_ns(&pkt->timestamp);

		/* Convert the network-wide distributed syntonized time
		 * representation to a timepoint as represented by the local
		 * interface's RAT.
		 */
		if (net_time_reference_get_timepoint_from_time(
			    ieee802154_cc13xx_cc26xx_net_time_reference_api_get(), tx_time,
			    NET_TIME_ROUNDING_NEAREST_TIMEPOINT, &tx_timepoint)) {
			return -EIO;
		}

		/* Convert the opaque counter timepoint into a TI driver specific
		 * RAT value.
		 */
		net_time_counter_get_tick_from_timepoint(
			ieee802154_cc13xx_cc26xx_net_time_reference_api_get(), tx_timepoint,
			&rat_tx_time);

		drv_data->cmd_prop_tx_adv.startTrigger.triggerType = TRIG_ABSTIME;
		drv_data->cmd_prop_tx_adv.startTime = rat_tx_time;
	}
```

The abstraction levels are:
* vendor/driver specific RAT ticks: ratmr_t,
* abstract/opaque counter ticks: k_timepoint_t,
* distributed syntonized time: net_time_t

Commit:
kernel: timeout: special timepoint values

Introduces definitions for special timepoints corresponding to a
timepoint that can "never" occur and a "zero" timepoint.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
drivers: ieee802154: introduce network uptime API

Optionally introduces the network uptime API into the IEEE 802.15.4
driver API.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
kernel: sys_clock: concentrate state in struct

Prepare for instantiable system clock by concentrating its state in a
struct.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

fgrandel commented:
This whole PR is still mainly blocked by #58080. @stephanosio - Can we get instructions how to reliably work around the issue until a fix can be provided? Any small change to memory layout seems to break the build.

stephanosio commented:
> Since its inception some three months ago, this whole PR is still mainly blocked by #58080. @stephanosio - Can we get instructions how to reliably work around the issue until a fix can be provided? Any small change to memory layout currently breaks random build tasks. I know there are other issues but it doesn't really make sense to fine tune them unless I can rely on automated tests to signal more than noise.

@fgrandel I just recently came back to work and am currently going through a lengthy TODO list. I will try to release a new SDK version with the fix proposed in https://github.com/zephyrproject-rtos/zephyr/issues/58080#issuecomment-1555163828 as soon as possible (likely in two weeks time).

fgrandel commented:
> likely in two weeks time

@stephanosio This is great news. :-) Thank you very much for coming back to me - it feels much better to know that the issue is not forgotten.

Commit:
[DNM] arm: linker: workaround for #58080

Padding section alignment for 64bit-aligned timeout struct.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
kernel: timeout: generalize timeout api

Adds the elapsed() and set_timeout() operations to the timeout api and
decouples all calls to these operations from the system clock.

This prepares for future pluggability of these operation.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
tests: kernel: pipe: increase stack size

Minimally increase stack size on the kernel pipe test as the additional
indirection required by the introduction of a pluggable kernel timer API
seems to exceed the required stack size just by a few bytes.

Also fixes a test isolation issue: One of the tests didn't prepare it's
own precondition which made it impossible to execute in isolation.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
kernel: timeout: separate sys_clock Kconfig from timeout api

Decouples the activation of the system clock from the activation of the
generic timeout API so that the latter can be enabled independently from
the former.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
kernel: timeout: publish multi-instance to api

Exposes the newly created multi-instance timeout API via timeout_q.h
header file.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: kernel: clocks: fix macro references

The recent refactoring of time conversion functions to macros left a few
references orphaned in the docs.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: kernel: clock: prefix system clock apis

Now that "clocks" can mean system clocks or abstract clocks, the "clock"
concept has become ambivalent. Also there is an inconsistency between
system clock header files (that use the sys_ prefix) and docs (that do
not use the prefix).

Both, the ambivalence and the inconsistency, are being fixed by
prefixing the system clock doc group with "sys".

The change also renames "timeutil_unit_apis" which is also inconsistent
and not really readable anyway.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
net: time: minor doc improvement

Updates the existing net time docs to refer to the recently introduced
timepoint concept to further clarify what is meant by tick counter
values.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
kernel: time: header to represent generic time concepts

The header files `zephyr/sys/time_units.h` and `zephyr/sys_clock.h` had
both, generic time conversion and system clock specific utilities. This
change creates a header file that will provide generic utilities to
represent and convert time to be used independently from the system
clock.

The newly created `zephyr/sys_time.h` header must not depend on any
other header files except for low level generic type declarations, the
toolchain and low level utilities so that it can be included from
`time_units.h` and `sys_clock.h` without include loops.

The header is named `zephyr/sys_time.h` to avoid confusion with the
POSIX `sys/time.h` header. The chosen name is consistent with existing
sys_* headers.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: sys: time_units: let doxygen recognize doc target

Fixes a minor bug that prevented doxygen from correctly recognizing the
function to which the doc comment was attached.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys: time_units: move time to timeout conversion

Moves generic time to timeout conversion utils to the general time
header.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys: time_units: move time converter generator macros

Moves generic time converter generator macros to the general time
header.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: sys_time: slightly improve time converter macro docs

Slightly improves generic time converter macro documentation.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys: time_units: remove obsolete TIME_CONSTEXPR

The TIME_CONSTEXPR definition was obsoleted by
d8b276e929c73690462ca05068330748093bfff0 and can be removed now.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
doc: sys: time_units: improve readability

Improve readability of inline documentation by re-ordering related
comments.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys_clock: move public generic time constants

Moves public generic time constants that are not specific to the system
clock into the generic time header.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys_clock: move private generic time constants

Moves private generic time constants that are not specific to the system
clock into the generic time header.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys_clock: move timepoint related utilities

Moves all time point related utilities to the generic time utilities as
these are not specifically bound to the system clock.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>

Commit:
sys_clock: clean up condition macros

Now that the content of the sys_clock header is reduced to what is
really specific to the system clock, the condition macros and
documentation structure can be simplified and cleaned up.

Signed-off-by: Florian Grandel <fgrandel@code-for-humans.de>
